### PR TITLE
Python: implemented web search engine skill with bing connector

### PR DESCRIPTION
**Pull request by Sailesh-2209**
### Motivation and Context

In this PR, I have tried my hand at an implementation of web search engine skill in python semantic kernel using the Bing Web Search API.

### Description

In the semantic kernel directory, I have added a new directory called web_skills (To replicate Skills.Web from C#) and added the web search skill here. For now, I have implemented web search using the bing web search API. If this approach is fine, then I can implement the same with the google search API too. I have tried to stick with similar naming conventions as used in the C# implementation with matching context parameters and arguments.

I can also add some unit tests for the connectors and the search skill, and add something like exponential backoff to avoid rate limit errors while querying the search APIs.

Here is some sample code that checks the working of the search skill.

```python
import os
import semantic_kernel as sk
from semantic_kernel.web_skills.web_search_engine_skill import WebSearchEngineSkill
from semantic_kernel.web_skills.connectors import BingConnector
from semantic_kernel.connectors.ai.open_ai import OpenAITextCompletion

async def main():
    kernel = sk.Kernel()
    api_key, org_id = sk.openai_settings_from_dot_env()
    kernel.add_text_completion_service(
        "dv", OpenAITextCompletion("text-davinci-003", api_key, org_id)
    )
    connector = BingConnector(api_key=os.getenv("BING_API_KEY"))
    web_skill = kernel.import_skill(WebSearchEngineSkill(connector), "WebSearch")

    prompt = "Who is Leonardo DiCaprio's current girlfriend?"
    search_async = web_skill["searchAsync"]
    result = await search_async.invoke_async(prompt)
    print(result)

    """
    Output:
    ["Celebrity Celebrity News Everything You Need to Know About Leonardo DiCaprio and Camila Morrone's Relationship From the beginning of their romance to today, we track their relationship here. By..."]
    """

    prompt = """
    Answer the question using only the data that is provided in the data section. Do not use any prior knowledge to answer the question.
    Data: {{WebSearch.SearchAsync "What is semantic kernel?"}}
    Question: What is semantic kernel?
    Answer:
    """

    qna = kernel.create_semantic_function(prompt, temperature=0.2)
    context = kernel.create_new_context()
    context["count"] = "10"
    context["offset"] = "0"
    result = await qna.invoke_async(context=context)
    print(result)

    """
    Output:
    Semantic Kernel is an open-source SDK that lets you easily combine AI services like OpenAI, Azure OpenAI, and Hugging Face with conventional programming languages like C# and Python. By doing so, you can create AI apps that combine the best of both worlds. Semantic Kernel is at the center of the copilot stack.
    """

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
```

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/web_skills/web_search_engine_skill.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from semantic_kernel.web_skills.connectors import Connector
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/web_search_engine_skill.py:1]:
```suggestion
from semantic_kernel.web_skills.connectors import ConnectorBase
```

File: python/semantic_kernel/web_skills/web_search_engine_skill.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    _connector: Connector
+
+    def __init__(self, connector: Connector) -> None:
+        self._connector = connector
+
+    @sk_function(
+        description="Performs a web search for a given query", name="searchAsync"
+    )
+    @sk_function_context_parameter(
+        name="count",
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/web_search_engine_skill.py:27]:
```suggestion
        name="num_results",
```

File: python/semantic_kernel/web_skills/web_search_engine_skill.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        """
+        Returns the search results of the query provided.
+        Returns `count` results and ignores the first `offset`.
+
+        :param query: search query
+        :param context: contains the context of count and offset parameters
+        :return: stringified list of search results
+        """
+
+        _, _count = context.variables.get("count")
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/web_search_engine_skill.py:46]:
```suggestion
        _num_results= context.variables["num_results"]
```

File: python/semantic_kernel/web_skills/web_search_engine_skill.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        _, _offset = context.variables.get("offset")
+        result = await self._connector.search_async(query, _count, _offset)
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/web_search_engine_skill.py:48]:
```suggestion
        result = await self._connector.search_async(query, _num_results, _offset)
```

File: python/semantic_kernel/web_skills/connectors/bing_connector.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        Returns `count` results and ignores the first `offset`.
+
+        :param query: search query
+        :param context: contains the context of count and offset parameters
+        :return: list of search results
+        """
+        if not query:
+            raise ValueError("query cannot be 'None' or empty.")
+
+        if not count:
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/connectors/bing_connector.py:32]:
```suggestion
        if not num_results:
```

File: python/semantic_kernel/web_skills/connectors/bing_connector.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            count = 1
+        if not offset:
+            offset = 0
+
+        count = int(count)
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/connectors/bing_connector.py:37]:
```suggestion
        num_results = int(num_results)
```

File: python/semantic_kernel/web_skills/connectors/bing_connector.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        if count <= 0:
+            raise ValueError("count value must be greater than 0.")
+        if count >= 50:
+            raise ValueError("count value must be less than 50.")
+
+        if offset < 0:
+            raise ValueError("offset must be greater than 0.")
+
+        self._logger.info(
+            f"Recieved request for bing web search with params:\nquery: {query}\ncount: {count}\noffset: {offset}"
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/connectors/bing_connector.py:49]:
```suggestion
            f"Recieved request for bing web search with params:\nquery: {query}\nnum_results: {num_results}\noffset: {offset}"
```

File: python/semantic_kernel/web_skills/connectors/bing_connector.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        )
+
+        _base_url = "https://api.bing.microsoft.com/v7.0/search"
+        _request_url = f"{_base_url}?q={urllib.parse.quote_plus(query)}&count={count}&offset={offset}"
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/connectors/bing_connector.py:53]:
```suggestion
        _request_url = f"{_base_url}?q={urllib.parse.quote_plus(query)}&count={num_results}&offset={offset}"
```

File: python/semantic_kernel/web_skills/connectors/connector.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class Connector:
+    """
+    Base class for search engine connectors
+    """
+
+    def search_async(self, query: str, count: int, offset: int) -> str:
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/connectors/connector.py:6]:
```suggestion
    def search_async(self, query: str, num_results: int, offset: int) -> str:
```

File: python/semantic_kernel/web_skills/web_search_engine_skill.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+class WebSearchEngineSkill:
+    """
+    Description: A skill that provides web search engine functionality
+
+    Usage:
+        connector = BingConnector(bing_search_api_key)
+        kernel.import_skill(WebSearchEngineSkill(connector), skill_name="WebSearch")
+
+    Examples:
+        {{WebSearch.SearchAsync "What is semantic kernel?"}}         => Returns the first `count` number of results for the given search query and ignores the first `offset` number of results (count and offset are specified in SKContext)
```
Comment (awharrison-28) [inline python/semantic_kernel/web_skills/web_search_engine_skill.py:15]:
```suggestion
        {{WebSearch.SearchAsync "What is semantic kernel?"}}         => Returns the first `num_results` number of results for the given search query and ignores the first `offset` number of results (num_results and offset are specified in SKContext)
```

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
- I like the example code that you put in the PR description. It'd be great if you could add that to `samples/python/kernel-syntax-examples`
- `Connector -> ConnectorBase` for consistency with SK patterns
- Suggestion: rename `count` to `num_results` (I've made it easy to accept if you agree)
- Runing `poetry run pre-commit run -c .conf/.pre-commit-config.yaml -a` should fix your CI errors
- when you feel that you've addressed the feedback, please change the label from **PR: feedback to address** to **PR: ready for review** so to let the team know :)

**Comment by Sailesh-2209** [comment]
@awharrison-28 I've addressed all the feedback and made the necessary changes. In `web_search_engine_skill.py`, I've kept the syntax for accessing context variables the same (`_, _num_results = context.variables.get("num_results")`) because it was causing an error when either one of num_results or offset parameters weren't being passed. 

I don't think I am allowed to change the PR label to **PR: ready for review** though

**Review by alexchaomander — APPROVED** [review]
Looks good to me!

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
Thanks for adding the sample. A couple of things: 

1. When running the sample, I'm getting an error. Can you check if you can repro? :

> Error: Cannot serialize non-str key None
> Error: Function `WebSearch.SearchAsync` execution failed. TypeError: Cannot serialize non-str key None

2. For organization, I think it makes more sense to move `semantic_kernel/web_skills/connectors/*` to `python/semantic_kernel/connectors` and `semantic_kernel/web_skills/web_search_engine_skill.py` to `semantic_kernel/core_skills/web_search_engine_skill.py`

**Comment by Sailesh-2209** [comment]
> Error: Cannot serialize non-str key None
> Error: Function WebSearch.SearchAsync execution failed. TypeError: Cannot serialize non-str key None
- This above error is caused when the environment variable BING_API_KEY is missing (or not present in `.env`). I now throw an appropriate error when the key is missing.
- Fixed the circular import issue.
- The directory `python/semantic_kernel/connectors` already exists and has two sub-directories - `ai` and `memory`. Shall I create another directory called `web` and move the connectors there?

**Comment by awharrison-28** [comment]
> > Error: Cannot serialize non-str key None
> > Error: Function WebSearch.SearchAsync execution failed. TypeError: Cannot serialize non-str key None
> 
> * This above error is caused when the environment variable BING_API_KEY is missing (or not present in `.env`). I now throw an appropriate error when the key is missing.
> * Fixed the circular import issue.
> * The directory `python/semantic_kernel/connectors` already exists and has two sub-directories - `ai` and `memory`. Shall I create another directory called `web` and move the connectors there?

I suggest calling the directory `search_engine`

**Comment by Sailesh-2209** [comment]
> Hmm I understand the error, but I was getting it when I had an .env variable for BING_API_KEY set.

Oh sorry about that, I'd forgotten to load env variables from `.env`, I've now added a call to `load_dotenv`. If there is a better way to do this, I could change it in the example.

**Review by awharrison-28 — APPROVED** [review]
Awesome work. Thank you for the contribution :)

**Review by awharrison-28 — APPROVED** [review]

**Pull request merged**